File: languages/go/entity/encryption/rsa_encrypter.go
package encryption

import (
	"github.com/bloock/bloock-sdk-go/v2/entity/key"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
)

type RsaEncrypter struct {
	Alg  proto.EncryptionAlg
	Args EncrypterArgs
}

func NewRsaEncrypter(args EncrypterArgs) RsaEncrypter {
	return RsaEncrypter{
		Alg:  proto.EncryptionAlg_RSA,
		Args: args,
	}
}

func (s RsaEncrypter) ToProto() *proto.Encrypter {
	var localKey *proto.LocalKey
	if s.Args.LocalKey != nil {
		localKey = s.Args.LocalKey.ToProto()
	} else if s.Args.Key != "" {
		localKey = key.LocalKey{
			Key: s.Args.Key,
		}.ToProto()
	}

	var managedKey *proto.ManagedKey
	if s.Args.ManagedKey != nil {
		managedKey = s.Args.ManagedKey.ToProto()
	}

	return &proto.Encrypter{
		Alg:        s.Alg,
		LocalKey:   localKey,
		ManagedKey: managedKey,
	}
}

File: languages/js/src/entity/encryption/rsa_encrypter.ts
import * as proto from "../../bridge/proto/encryption_entities";
import { KeyType, LocalKey, ManagedKey } from "../key";
import { Encrypter } from "./encrypter";
import { EncrypterArgs } from "./encrypter_args";

export class RsaEncrypter implements Encrypter {
  alg: proto.EncryptionAlg;
  args: EncrypterArgs;

  constructor(key: LocalKey | ManagedKey | string) {
    this.alg = proto.EncryptionAlg.RSA;

    let encrypter_key;
    if (key instanceof LocalKey) {
      encrypter_key = key;
    } else if (key instanceof ManagedKey) {
      encrypter_key = key;
    } else {
      encrypter_key = new LocalKey(key, KeyType.Rsa2048);
    }
    this.args = new EncrypterArgs(encrypter_key);
  }

  public toProto(): proto.Encrypter {
    return proto.Encrypter.fromPartial({
      alg: this.alg,
      localKey: this.args.localKey?.toProto(),
      managedKey: this.args.managedKey?.toProto()
    });
  }
}
